pr-deploy/test/app.test.ts

```diff
@@ -17,115 +17,138 @@
+      ),
   };
 });
 
-import {Application} from 'probot';
+process.env.GH_CHECK = 'test-check';
+process.env.GH_OWNER = 'test-owner';
+process.env.GH_REPO = 'test-repo';
+
+import nock from 'nock';
 import prDeployAppFn from '../src/app';
-import express from 'express';
-import request from 'supertest';
-import Webhooks, {
-  WebhookPayloadPullRequest,
-  WebhookPayloadCheckRun} from '@octokit/webhooks';
+import Webhooks from '@octokit/webhooks';
+import {Probot, ProbotOctokit} from 'probot';
+
+const apiUrl = 'https://api.github.com';
 
 describe('test pr deploy app', () => {
-  let app: Application;
-  let github: any;
-  let server: express.Application;
+  let probot: Probot;
 
   beforeEach(() => {
-    process.env.DISABLE_WEBHOOK_EVENT_CHECK = 'true';
-
-    github = {
-      checks: {
-        create: jest.fn().mockReturnValue(Promise.resolve()),
-        update: jest.fn().mockReturnValue(Promise.resolve()),
-        listForRef: jest.fn(),
-      },
-    };
-
-    app = new Application();
-    app.load(prDeployAppFn);
-    app.auth = async() => Promise.resolve(github);
+    nock.disableNetConnect();
+    probot = new Probot({
+      id: 1,
+      githubToken: 'test',
+      // Disable throttling & retrying requests for easier testing
+      Octokit: ProbotOctokit.defaults({
+        retry: {enabled: false},
+        throttle: {enabled: false},
+      }),
+    });
+    probot.load(prDeployAppFn);
+  });
 
-    server = express();
-    server.use(app.router);
+  afterEach(() => {
+    expect(nock.isDone()).toBeTruthy();
+    nock.cleanAll();
+    nock.enableNetConnect();
   });
 
   test('creates a check when a pull request is opened', async() => {
-    // make sure no checks already exist
-    github.checks.listForRef.mockReturnValue(null);
-
-    const prOpenedEvent: Webhooks.WebhookEvent<WebhookPayloadPullRequest> = {
+    const prOpenedEvent: Webhooks.WebhookEvent<
+    Webhooks.EventPayloads.WebhookPayloadPullRequest> = {
       id: 'prId',
       name: 'pull_request.opened',
       payload: {
         pull_request: {head: {sha: 'abcde'}},
-        repository: {owner: {name: 'repoOwner'},name: 'newRepo'},
-      } as WebhookPayloadPullRequest,
+        repository: {
+          owner: {name: 'test-owner'},
+          name: 'test-repo',
+        },
+      } as Webhooks.EventPayloads.WebhookPayloadPullRequest,
     };
 
-    await app.receive(prOpenedEvent);
-    expect(github.checks.create).toHaveBeenCalled();
-    expect(github.checks.update).not.toHaveBeenCalled();
+    nock(apiUrl)
+      .get('/repos/test-owner/test-repo/commits/' +
+      'abcde/check-runs?check_name=test-check')
+      .reply(200, null) // make sure no checks already exist
+      .post('/repos/test-owner/test-repo/check-runs', body => {
+        expect(body).toMatchObject({
+          'head_sha': 'abcde',
+          'name': 'test-check',
+          'status': 'queued',
+        });
+        return true;
+      })
+      .reply(200);
```

Comment:
Oh no 😱 and just when I did the exact opposite (#1005, #1009)

I'm gonna argue the exact opposite! Octokit is the thing you should use for expectations, the *network calls* are the internal implementation detail. Think about it this way - when you write code, do you write `probot.networkCall('/repos/test-owner/test-repo/commits/abcde/check-runs?check_name=test-check', payload)` or do you write `probot.github.checks.create(payload)`?

I hated mocking the network calls when I first wrote the tests because that's what felt wrong about it, when I saw how you solved it I was super happy... now I'm sad 😭

Reply:
I lean heavily in support of Nock for testing GitHub API calls. There are cases where probot or octokit make calls, or in some cases requests are even built manually for the rare thing octokit doesn't do. The API format is the static, external piece that we want to ensure is called, and that rarely changes and is well documented. Mocking out the various calls of the (inconsistent and often changing) octokit/probot libraries seems more fragile to me and closer to an implementation detail, especially when different apps sometimes are on different versions of each library.

For example, something fetching multiple pages of [insert resource here] might do it manually, or might use octokits auto-pagination feature. They would require entirely different test mocking, but the underlying API calls should be the same.

Reply:
I agree with Ryan, because network calls are further from Probot's internal implementation, whereas Octokit is an internal dependency. So if Probot decides to use its own GH client, our tests will not break. Intercepting http requests is what Probot docs suggest as well.

That said, if Octokit decided to use different queries, like graphql instead of REST, then our tests would break. But I'd imagine this would come with a major version upgrade.

And personally I liked mocking Octokit because it reads a lot nicer in tests... imo both ways of testing are fine!